### PR TITLE
Compression: Refactor to support multiple compression

### PR DESCRIPTION
**Pull request by Lloyd-Pottiger**
### What problem does this PR solve?

Issue Number: ref #8789

Problem Summary:

### What is changed and how it works?

<del> https://github.com/pingcap/tiflash/pull/8783 should merge first, the diff is https://github.com/Lloyd-Pottiger/tiflash/compare/refactor-compression...Lloyd-Pottiger:tiflash:support-multiple-compression <del>

Introduce a interface `ICompressionCodec` which represents interface for compression codecs like LZ4, ZSTD, etc.

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/CMakeLists.txt [hunk 1]
```diff
@@ -179,6 +179,7 @@ target_link_libraries (tiflash_common_io
     ${LINK_LIBRARIES_ONLY_ON_X86_64}
     ${LZ4_LIBRARY}
     ${ZSTD_LIBRARY}
+    ${QPL_LIBRARY}
```
Comment (JaySon-Huang) [inline dbms/CMakeLists.txt:182]: this line should be put to
```
if (USE_QPL)
    target_link_libraries(tiflash_common_io ${QPL_LIBRARY})
... 
```

File: dbms/CMakeLists.txt [hunk 2]
```diff
@@ -262,8 +263,10 @@ if (NOT USE_INTERNAL_ZSTD_LIBRARY)
 endif ()
 
 if (USE_QPL)
-    target_link_libraries (dbms ${QPL_LIBRARY})
-    target_include_directories (dbms PRIVATE ${QPL_INCLUDE_DIR})
+    target_include_directories (dbms BEFORE PRIVATE ${QPL_INCLUDE_DIR})
+    set_source_files_properties(
+        src/IO/Compression/CompressionCodecDeflateQpl.cpp
+        PROPERTIES COMPILE_FLAGS "-mwaitpkg")
```
Comment (JaySon-Huang) [inline dbms/CMakeLists.txt:269]: What does this compile_flags mean? Why is this required now but not in the previous build?

Comment (Lloyd-Pottiger) [reply dbms/CMakeLists.txt:269]:
```
_tpause(1, __rdtsc() + 1000);
```

https://www.felixcloutier.com/x86/tpause

File: dbms/CMakeLists.txt (hunk 1 above)
Comment (Lloyd-Pottiger) [reply dbms/CMakeLists.txt:182]: When USE_QPL = 0, ${QPL_LIBRARY} is empty, I think there is no need to add if branch.

File: dbms/src/IO/Compression/CompressedReadBufferBase.cpp [hunk 3]
```diff
@@ -43,6 +33,56 @@ extern const int CANNOT_DECOMPRESS;
...
+    size_t & size_decompressed,
+    size_t & size_compressed_without_checksum)
+{
+    readHeaderAndGetCodec(compressed_buffer, codec);
+
+    size_compressed_without_checksum = codec->readCompressedBlockSize(compressed_buffer);
+    size_decompressed = codec->readDecompressedBlockSize(compressed_buffer);
+
+    /// This is for clang static analyzer.
+    assert(size_decompressed > 0);
```
[the remark this answers is not in the document]
Comment (Lloyd-Pottiger) [reply dbms/src/IO/Compression/CompressedReadBufferBase.cpp:68]: `size_decompressed` will always be greater 0.

[the remark this answers is not in the document]
Comment (CalvinNeo) [reply dbms/src/IO/Compression/CompressedReadBufferBase.cpp:68]: OK, I get that point

File: dbms/CMakeLists.txt (hunk 1 above)
Comment (JaySon-Huang) [reply dbms/CMakeLists.txt:182]: I think put it to the if branch is more clear

File: dbms/src/IO/Compression/CompressionInfo.h [hunk 4]
```diff
@@ -14,33 +14,18 @@
...
-    ZSTD = 3, /// Experimental algorithm: https://github.com/Cyan4973/zstd
-#if USE_QPL
-    QPL = 4, /// The Intel Query Processing Library (QPL) is an open-source library to provide high-performance query processing operations
-    NONE = 5, /// No compression
-#else
-    NONE = 4, /// No compression
-#endif
-};
-
+using UInt8 = std::uint8_t;
```
Comment (JaySon-Huang) [inline dbms/src/IO/Compression/CompressionInfo.h:28]: Use `#include <common/types.h>` to avoid defining another `DB::UInt8` rather thant `::UInt8`?

File: dbms/CMakeLists.txt (hunk 1 above)
Comment (Lloyd-Pottiger) [reply dbms/CMakeLists.txt]: Remove this line, and 

```
if (USE_QPL)
    target_link_libraries (tiflash_common_io PUBLIC ${QPL_LIBRARY}) <<--- add here
    target_include_directories (dbms BEFORE PRIVATE ${QPL_INCLUDE_DIR})
    set_source_files_properties (
        src/IO/Compression/CompressionCodecDeflateQpl.cpp
        PROPERTIES COMPILE_FLAGS "-mwaitpkg")
endif ()
```

File: dbms/src/IO/Compression/CompressionInfo.h (hunk 4 above)
Comment (Lloyd-Pottiger) [reply dbms/src/IO/Compression/CompressionInfo.h]: done

File: dbms/src/IO/Compression/CompressionFactory.h [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+}
+
+class CompressionFactory
+{
+public:
+    static CompressionCodecPtr create(const CompressionSettings & settings)
+    {
+        switch (settings.method)
+        {
+        case CompressionMethod::LZ4:
```
[the remark this answers is not in the document]
Comment (Lloyd-Pottiger) [reply dbms/src/IO/Compression/CompressionFactory.h:42]: no, We may apply different compression algorithms for different data.

File: dbms/src/IO/Compression/CompressionFactory.h [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+                ErrorCodes::UNKNOWN_COMPRESSION_METHOD,
+                "Unknown compression method: {}",
+                static_cast<int>(settings.method));
+        }
+    }
+
+    static CompressionCodecPtr create(UInt8 method_byte)
+    {
+        CompressionSettings settings;
+        switch (method_byte)
```
[the remark this answers is not in the document]
Comment (Lloyd-Pottiger) [reply dbms/src/IO/Compression/CompressionFactory.h]: Good idea, addressed.

File: dbms/src/IO/Compression/CompressionFactory.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (Lloyd-Pottiger) [reply dbms/src/IO/Compression/CompressionFactory.h:44]: I plan to add a compression algorithm "delta", which behaves differently when the column type is `int8/int16/int32/int64`, so this algorithm is not stateless and can not use one global object.

[the remark this answers is not in the document]
Comment (Lloyd-Pottiger) [reply dbms/src/IO/Compression/CompressionFactory.h:44]: I found the config `dt_compression_level` needs to support hot load. So even LZ4 and ZSTD is not stateless.

File: libs/libcommon/include/common/unaligned.h [hunk 7]
```diff
@@ -16,12 +16,14 @@
...
 #include <string.h>
 
+#include <bit>
 #include <type_traits>
 
 
 template <typename T>
 inline T unalignedLoad(const void * address)
 {
+    static_assert(std::endian::native == std::endian::little, "Only little endian is supported");
```
Comment (JaySon-Huang) [inline libs/libcommon/include/common/unaligned.h:26]: `static_assert` only ensure the compile machine is little endian.
Maybe we need to add a explicit check when tiflash starts. And do that in a standalone PR

Comment (Lloyd-Pottiger) [reply libs/libcommon/include/common/unaligned.h]: reverted

File: dbms/src/IO/Compression/ICompressionCodec.h [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+    static UInt8 readMethod(const char * source);
+
+    /// Return true if this codec actually compressing something. Otherwise it can be just transformation that helps compression (e.g. Delta).
+    virtual bool isCompression() const = 0;
+
+    /// Is it a generic compression algorithm like lz4, zstd. Usually it does not make sense to apply generic compression more than single time.
+    virtual bool isGenericCompression() const = 0;
+
+    /// Is this the DEFLATE_QPL codec?
+    virtual bool isDeflateQpl() const { return false; }
```
Comment (JaySon-Huang) [inline dbms/src/IO/Compression/ICompressionCodec.h:68]: why we need this special method for deflatQpl? Can we use `getMethodByte() == CompressionMethodByte::QPL` instead?

Comment (Lloyd-Pottiger) [reply dbms/src/IO/Compression/ICompressionCodec.h]: Currently not used, removed.

File: dbms/src/IO/Compression/ICompressionCodec.h [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+class ICompressionCodec : private boost::noncopyable
+{
+public:
+    virtual ~ICompressionCodec() = default;
+
+    /// Byte which indicates codec in compressed file
+    virtual UInt8 getMethodByte() const = 0;
+
+    /// Compressed bytes from uncompressed source to dest. Dest should preallocate memory
+    UInt32 compress(const char * source, UInt32 source_size, char * dest) const;
```
[the remark this answers is not in the document]
Comment (Lloyd-Pottiger) [reply dbms/src/IO/Compression/ICompressionCodec.h:38]: `dest_size` is unknown before compress.

File: dbms/CMakeLists.txt (hunk 2 above)
Comment (Lloyd-Pottiger) [reply dbms/CMakeLists.txt:270]: https://github.com/pingcap/tiflash/blob/5c2d39d6fb2c7eb55d28f2d0495a1a26f5030d23/dbms/src/IO/Compression/CompressionCodecDeflateQpl.cpp#L239

here

File: dbms/src/IO/Compression/ICompressionCodec.h (hunk 9 above)
[the remark this answers is not in the document]
Comment (Lloyd-Pottiger) [reply dbms/src/IO/Compression/ICompressionCodec.h:38]: Each algorithm should overload the `getMaxCompressedDataSize` method.
```
UInt32 CompressionCodecLZ4::getMaxCompressedDataSize(UInt32 uncompressed_size) const
```
